Fix type of indirect function calls

in C mode we should apply a function to pointer conversion on the base expression and fix the type checker to handle the call.

```c
int (*fptr)();

int main() {
  _Generic(fptr(), int: 1);
  return 0;
}
```

```bash
$ cxx x.c -fcheck

x.c:4:3: error: generic selection expression with invalid selector type
  _Generic(fptr(), int: 1);
  ^
```

--- a/src/parser/cxx/type_checker.cc
+++ b/src/parser/cxx/type_checker.cc
@@ -460,9 +460,31 @@ void TypeChecker::Visitor::operator()(CallExpressionAST* ast) {
   }
 
   auto functionType = type_cast<FunctionType>(ast->baseExpression->type);
+
   if (!functionType) {
+    if (control()->is_pointer(ast->baseExpression->type)) {
+      // ressolve pointer to function type
+      functionType = type_cast<FunctionType>(
+          control()->get_element_type(ast->baseExpression->type));
+    }
+
+    if (functionType && is_parsing_c()) {
+      (void)ensure_prvalue(ast->baseExpression);
+    }
+  }
+
+  if (!functionType) {
+    // todo: enable when support for the __builtin_<op> functions is added
+
+#if false
+    error(ast->firstSourceLocation(),
+          std::format("invalid call of type '{}'",
+                      to_string(ast->baseExpression->type)));
+#endif
     return;
   }
+
+  // TODO: check the arguments
 
   ast->type = functionType->returnType();
 

--- a/tests/unit_tests/sema/call_c_01.c
+++ b/tests/unit_tests/sema/call_c_01.c
@@ -1,0 +1,8 @@
+// RUN: %cxx -verify -fcheck %s
+
+int f() { return 0; }
+
+int main() {
+  int (*fptr)() = f;
+  (void)_Generic(fptr(), int: 1);
+}

